packages/promise-queue/src/PromiseQueue.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Wire
+ * Copyright (C) 2022 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {PromiseQueueOptions} from './PromiseQueueOptions';
+import {QueueEntry, Task} from './QueueEntry';
+
+const defaultOptions = {
+  timeout: 1000 * 60, // 1 minute
+  concurrent: 1,
+  paused: false,
+};
+
+export class PromiseQueue {
+  private blocked: boolean;
+  private runningTasks: number;
+  private interval?: number;
+  private paused: boolean;
+  private readonly concurrent: number;
+  private readonly logger?: {warn: (...args: any[]) => void};
```

Comment:
just curious if we should include a default logger option as well as allowing the consumer to set one.

Reply:
I'd like our libraries to be rather agnostic from any particular logger. 
Initially it pulled the entire `logdown` lib for logging, but I don't think it makes sense. If at some point we migrate the webapp's logger, I would like the change to reflect on all the libraries we use. 
So I think injecting the logger makes a lot of sense and let the webapp completely control how things are logged and with what namespace